### PR TITLE
Add probe for Kubernetes cron_jobs

### LINKED ISSUE #27: Add probe for Kubernetes cron jobs
Add Kubernetes probe for [Cron Job](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) resource. Cron job elements should be linked to job elements.

Python client method: [list_namespaced_cron_job](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/BatchV1beta1Api.md#list_namespaced_cron_job) (see [all methods](https://github.com/kubernetes-client/python/blob/master/kubernetes/README.md#documentation-for-api-endpoints))

In order to add a new probe, one need to implement **probe** and **linker**. The former pulls upstream entities (i.e. K8S objects) and inserts them into entity graph. The latter links inserted graph nodes. In other words: probes create graph nodes, whereas linkers create graph edges. 

Below, an example implementation of related modules for K8S Service object:

* [probe](https://github.com/openrca/orca/blob/3dcf1c957653e52ce34bb1bdc334ffca3508cb9e/orca/topology/infra/k8s/probe.py#L67) - provisions a probe instance
* [extractor](https://github.com/openrca/orca/blob/3dcf1c957653e52ce34bb1bdc334ffca3508cb9e/orca/topology/infra/k8s/extractor.py#L108) - extracts node element from raw upstream entity
* [linker](https://github.com/openrca/orca/blob/3dcf1c957653e52ce34bb1bdc334ffca3508cb9e/orca/topology/infra/k8s/linker.py#L24) - provisions a linker instance
* [matcher](https://github.com/openrca/orca/blob/3dcf1c957653e52ce34bb1bdc334ffca3508cb9e/orca/topology/infra/k8s/matcher.py#L23) - matches two nodes based on matching condition, used by linker
* [init](https://github.com/openrca/orca/blob/3dcf1c957653e52ce34bb1bdc334ffca3508cb9e/orca/topology/infra/k8s/__init__.py#L50) - provisions a probe bundle (probe + linkers)

### PR DESCRIPTION
**Pull request by aleksandra-galara**
It implements #27

Signed-off-by: Aleksandra Galara <a.galara@samsung.com>

### CODE REVIEW

File: orca/topology/infra/k8s/extractor.py [hunk 1]
```diff
@@ -353,3 +353,47 @@ def _extract_properties(self, entity):
...
+class CronJobExtractor(Extractor):
+
+    """Extractor for CronJob entities."""
+
+    @property
+    def kind(self):
+        return 'cron_job'
+
+    def _extract_properties(self, entity):
+        properties = {}
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/extractor.py:426]: Let's add `schedule` [property](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1beta1CronJobSpec.md) as well.

File: orca/topology/infra/k8s/extractor.py [hunk 2]
```diff
@@ -353,3 +353,47 @@ def _extract_properties(self, entity):
...
+        properties['name'] = entity.metadata.name
+        properties['namespace'] = entity.metadata.namespace
+        properties['uid'] = entity.metadata.uid
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/extractor.py]: The `uid` is [already extracted](https://github.com/openrca/orca/blob/master/orca/topology/infra/k8s/extractor.py#L35) for all K8S entities. You can access it by the `id` property of a graph node.

File: orca/topology/infra/k8s/matcher.py [hunk 3]
```diff
@@ -218,6 +218,22 @@ def are_linked(self, obj, namespace):
         return namespace.properties.name == obj.properties.namespace
 
 
+class CronJobToJobMatcher(Matcher):
+
+    """Generic matcher for links between CronJob and Job entities."""
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/matcher.py]: It's not a "generic" linker (e.g. for links between a namespace and ANY resource). It's a linker for links between concrete entity types: cron job and job. Remove the `generic` keyword from the docstring.

**Review by bzurkowski — CHANGES_REQUESTED** [review]
A few cosmetic remarks. Thanks!

Please, remember to rebase changes after completing #76. This PR should contain only changes related to cron jobs.

File: orca/topology/infra/k8s/linker.py [hunk 4]
```diff
@@ -254,3 +254,15 @@ def get(cls, graph):
             target_spec=utils.NodeSpec(origin='kubernetes', kind='pod'),
             matcher=matcher.JobToPodMatcher())
 
+
+class CronJobToJobLinker(Linker):
+
+    """Links CronJob entities to Job entities."""
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/linker.py:260]: Please, add one linebreak after the docstring.

File: orca/topology/infra/k8s/linker.py [hunk 5]
```diff
@@ -254,3 +254,15 @@ def get(cls, graph):
...
+     @classmethod
+     def get(cls, graph):
+         return cls(
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/linker.py]: The indent for these three lines is one space too far, causing the error:
```
05:17:14 probe.1 | Traceback (most recent call last):
05:17:14 probe.1 |   File "/usr/local/bin/orca-probe", line 11, in <module>
05:17:14 probe.1 |     load_entry_point('orca', 'console_scripts', 'orca-probe')()
05:17:14 probe.1 |   File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 489, in load_entry_point
05:17:14 probe.1 |     return get_distribution(dist).load_entry_point(group, name)
05:17:14 probe.1 |   File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2852, in load_entry_point
05:17:14 probe.1 |     return ep.load()
05:17:14 probe.1 |   File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2443, in load
05:17:14 probe.1 |     return self.resolve()
05:17:14 probe.1 |   File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2449, in resolve
05:17:14 probe.1 |     module = __import__(self.module_name, fromlist=['__name__'], level=0)
05:17:14 probe.1 |   File "/app/orca/cmd/probe.py", line 15, in <module>
05:17:14 probe.1 |     from orca.topology import manager
05:17:14 probe.1 |   File "/app/orca/topology/manager.py", line 22, in <module>
05:17:14 probe.1 |     from orca.topology import alerts, infra, probe
05:17:14 probe.1 |   File "/app/orca/topology/infra/__init__.py", line 15, in <module>
05:17:14 probe.1 |     from orca.topology.infra import istio, k8s, kiali
05:17:14 probe.1 |   File "/app/orca/topology/infra/istio/__init__.py", line 15, in <module>
05:17:14 probe.1 |     from orca.topology.infra.istio import linker, probe
05:17:14 probe.1 |   File "/app/orca/topology/infra/istio/probe.py", line 18, in <module>
05:17:14 probe.1 |     from orca.topology.infra.istio import extractor
05:17:14 probe.1 |   File "/app/orca/topology/infra/istio/extractor.py", line 17, in <module>
05:17:14 probe.1 |     from orca.topology.infra.k8s import extractor
05:17:14 probe.1 |   File "/app/orca/topology/infra/k8s/__init__.py", line 17, in <module>
05:17:14 probe.1 |     from orca.topology.infra.k8s import cluster, linker, probe
05:17:14 probe.1 |   File "/app/orca/topology/infra/k8s/linker.py", line 261
05:17:14 probe.1 |     @classmethod
05:17:14 probe.1 |     ^
05:17:14 probe.1 | IndentationError: unexpected indent
```

File: orca/topology/infra/k8s/matcher.py [hunk 6]
```diff
@@ -245,6 +245,23 @@ def are_linked(self, job, pod):
...
+    """Matcher for links between CronJob and Job entities."""
+
+    def are_linked(self, cron_job, job):
+        matched_namespace = match_namespace(cron_job, job)
+        matched_cron_job = self._match_cron_job(cron_job, job)
+        return matched_namespace and matched_cron_job
+
+    def _match_cron_job(self, cron_job, job):
+        if job.properties.owner_references:
+            for reference in job.properties.owner_references:
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/matcher.py]: Now it's`job.properties.cron_job_ref` 😉

**Review by bzurkowski — CHANGES_REQUESTED** [review]
I found some issues after the rebase. Please, fix them and we can merge.

File: orca/topology/infra/k8s/linker.py (hunk 5 above)
Comment (aleksandra-galara) [reply orca/topology/infra/k8s/linker.py]: I'm sorry for problems! Hope it's finally ok. ;)

File: orca/topology/infra/k8s/matcher.py [hunk 7]
```diff
@@ -245,6 +245,23 @@ def are_linked(self, job, pod):
...
+    """Matcher for links between CronJob and Job entities."""
+
+    def are_linked(self, cron_job, job):
+        matched_namespace = match_namespace(cron_job, job)
+        matched_cron_job = self._match_cron_job(cron_job, job)
+        return matched_namespace and matched_cron_job
+
+    def _match_cron_job(self, cron_job, job):
+        if job.properties.cron_job_ref:
+            for reference in job.properties.cron_job_ref:
```
Comment (bzurkowski) [inline orca/topology/infra/k8s/matcher.py]: One last remark. Remember, we changed `owner_references` (list) to `cron_job_ref` (single dict). In this case, you are iterating over keys in `cron_job_ref` dict:
```
06:39:26 probe.1 |   File "/app/orca/topology/infra/k8s/matcher.py", line 260, in _match_cron_job
06:39:26 probe.1 |     if reference.uid == cron_job.id:
06:39:26 probe.1 | AttributeError: 'str' object has no attribute 'uid'
```
I think it's time to introduce some basic testing to catch such small things. Recently, I catch myself on similar mistakes.

**Review by bzurkowski — CHANGES_REQUESTED** [review]
One last remark 😉

**Pull request merged**